Handle or forward

ashwanthkumar opened the pull request:
Fixes #2 

Early version of `HandleOrForward` based on a `ForwardStrategy`. Will integrate with Partitioner once #10 is merged. 

cc: @brewkode

ashwanthkumar committed:
WIP - Attempt at doing a HandleOrForward SeverInterceptor

ashwanthkumar committed:
Fixed the typo in the comment

ashwanthkumar committed:
Moving things around in rpc package

ashwanthkumar committed:
Add a note to handle forwarding loop

brewkode committed:
Merge branch 'master' into handle-or-forward

brewkode commented:
Making some changes towards better naming. Calling `ForwardStrategy` as `RoutingStrategy`. `handleOrForward` could be simply called `route`.

brewkode committed:
First cut CHP based Router

RequestForwarder => Router + other such renaming

ashwanthkumar committed:
Enabling DEBUG logs for all suuchi packages

ashwanthkumar committed:
Changing contract of RoutingStrategy#route -> PartialFunction

Also fixed a bug in the Router Interceptor when forwarding request

brewkode commented on src/main/scala/in/ashwanthkumar/suuchi/rpc/Router.scala:
```diff
@@ -0,0 +1,58 @@
+      override def onReady(): Unit = delegate.onReady()
+      override def onMessage(incomingRequest: ReqT): Unit = {
+        // TODO - Handle forwarding loop here
+        if(routingStrategy.route.isDefinedAt(incomingRequest)) {
```
Not too sure how this helps.

ashwanthkumar replied:
It helps when you don't want to forward certain kind of requests like health / ping. I know we can choose not to add this intercept for them, but I felt this also gives greater power in the hands of the developer. On the other hand it could also be deemed as confusing because of more than 2 ways to do the same thing. What do you think we should be doing?

brewkode replied:
Alright. We should then use `applyOrElse` as that seems a better way to do this (from the scaladoc)

> Note that expression `pf.applyOrElse(x, default)` is equivalent to {{{ if(pf isDefinedAt x) pf(x) else default(x) }}} except that `applyOrElse` method can be implemented more efficiently.

brewkode committed:
Code re-org + fixed Router

brewkode committed:
Fixing build failure caused during code re-org

brewkode committed:
CHR uses MemberAddress instead of Node

ashwanthkumar committed:
Adding default ConsistentHashRing object + minor refactoring

ashwanthkumar committed:
Removing empty lines in the file

ashwanthkumar committed:
Changing ServerInterceptors.intercept -> ServerInterceptors.interceptForward

This will apply the intercepts in FIFO

ashwanthkumar committed:
Yay! End to end integration of ConsistentHash based Partitioner on both the servers

ashwanthkumar committed:
Making sure we return only value Array[Byte] instead of key concat value

Fixed the SuuchiClient's main method to put A=A... Z=Z and do a corresponding get against them

ashwanthkumar committed:
Changing info -> trace messages

ashwanthkumar committed:
Removing the default un-used router

ashwanthkumar committed:
Merge branch 'master' into handle-or-forward

ashwanthkumar committed:
Removing maven-compiler plugin and doing incremental compile instead of modified-only in scala compiler plugin

ashwanthkumar committed:
Fixing warning in compilation for importing advanced language features

ashwanthkumar committed:
Adding happy path test cases in CHRSpec

ashwanthkumar committed:
happy path test cases for CHPartitioner

ashwanthkumar committed:
First cut tests for Router interceptor

ashwanthkumar commented:
I think we should merge this branch soon. I've experimented too much stuff in here. I'm very very happy with how things have progressed so far 👍

ashwanthkumar committed:
Doing log.debug -> log.trace
